scapy/contrib/isotp/isotp_native_socket.py

```diff
@@ -0,0 +1,368 @@
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more information
+# Copyright (C) Nils Weiss <nils@we155.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = ISO-TP (ISO 15765-2) Native Socket Library
+# scapy.contrib.status = library
+
+import ctypes
+from ctypes.util import find_library
+import struct
+import socket
+
+from scapy.compat import Optional, Union, Tuple, Type, cast
+from scapy.packet import Packet
+import scapy.modules.six as six
+from scapy.error import Scapy_Exception, warning
+from scapy.supersocket import SuperSocket
+from scapy.data import SO_TIMESTAMPNS
+from scapy.config import conf
+from scapy.arch.linux import get_last_packet_timestamp, SIOCGIFINDEX
+from scapy.contrib.isotp.isotp_packet import ISOTP
+from scapy.layers.can import CAN_MTU, CAN_MAX_DLEN
+
+
+LIBC = ctypes.cdll.LoadLibrary(find_library("c"))  # type: ignore
+
+CAN_ISOTP = 6  # ISO 15765-2 Transport Protocol
+
+SOL_CAN_BASE = 100  # from can.h
+SOL_CAN_ISOTP = SOL_CAN_BASE + CAN_ISOTP
+# /* for socket options affecting the socket (not the global system) */
+CAN_ISOTP_OPTS = 1  # /* pass struct can_isotp_options */
+CAN_ISOTP_RECV_FC = 2  # /* pass struct can_isotp_fc_options */
+
+# /* sockopts to force stmin timer values for protocol regression tests */
+CAN_ISOTP_TX_STMIN = 3  # /* pass __u32 value in nano secs    */
+CAN_ISOTP_RX_STMIN = 4  # /* pass __u32 value in nano secs   */
+CAN_ISOTP_LL_OPTS = 5  # /* pass struct can_isotp_ll_options */
+
+CAN_ISOTP_LISTEN_MODE = 0x001  # /* listen only (do not send FC) */
+CAN_ISOTP_EXTEND_ADDR = 0x002  # /* enable extended addressing */
+CAN_ISOTP_TX_PADDING = 0x004  # /* enable CAN frame padding tx path */
+CAN_ISOTP_RX_PADDING = 0x008  # /* enable CAN frame padding rx path */
+CAN_ISOTP_CHK_PAD_LEN = 0x010  # /* check received CAN frame padding */
+CAN_ISOTP_CHK_PAD_DATA = 0x020  # /* check received CAN frame padding */
+CAN_ISOTP_HALF_DUPLEX = 0x040  # /* half duplex error state handling */
+CAN_ISOTP_FORCE_TXSTMIN = 0x080  # /* ignore stmin from received FC */
+CAN_ISOTP_FORCE_RXSTMIN = 0x100  # /* ignore CFs depending on rx stmin */
+CAN_ISOTP_RX_EXT_ADDR = 0x200  # /* different rx extended addressing */
+
+# /* default values */
+CAN_ISOTP_DEFAULT_FLAGS = 0
+CAN_ISOTP_DEFAULT_EXT_ADDRESS = 0x00
+CAN_ISOTP_DEFAULT_PAD_CONTENT = 0xCC  # /* prevent bit-stuffing */
+CAN_ISOTP_DEFAULT_FRAME_TXTIME = 0
+CAN_ISOTP_DEFAULT_RECV_BS = 0
+CAN_ISOTP_DEFAULT_RECV_STMIN = 0x00
+CAN_ISOTP_DEFAULT_RECV_WFTMAX = 0
+CAN_ISOTP_DEFAULT_LL_MTU = CAN_MTU
+CAN_ISOTP_DEFAULT_LL_TX_DL = CAN_MAX_DLEN
+CAN_ISOTP_DEFAULT_LL_TX_FLAGS = 0
+
+
+class SOCKADDR(ctypes.Structure):
+    # See /usr/include/i386-linux-gnu/bits/socket.h for original struct
+    _fields_ = [("sa_family", ctypes.c_uint16),
+                ("sa_data", ctypes.c_char * 14)]
```

Comment:
This is far from standardized. See https://github.com/secdev/scapy/blob/100b2ca7d01128e059330d34f11d4b01d2744c7e/scapy/libs/winpcapy.py#L71

Reply:
I've updated the names. actually SOCKADDR wasn't used in my code.. Removed it! Thanks